### DISCUSSION TITLE
Clarification on usage

### DISCUSSION THREAD

**Discussion by zero-well** [euzu/tuliprox#40] [Q&A] [answered]
Context:  I have a rather large xtream iptv playlist (roughly 10k channels). I have been begrudgingly using Threadfin to filter out almost all of the channels. The way Threadfin works is I select a Group, and then select include/exclude statements and then the channels that adhere to those rules are kept. Here's an example:

![image](https://github.com/euzu/m3u-filter/assets/160749165/41bb7b86-87df-4057-b3e3-eecab872cc77)

The NFL League Pass group has 38 channels and I only want one channel named "Red Zone." It matches two channels though "Red Zone" and "Red Zone Low" so I set an exclusion rule for "low" and that filter adds that channel. 

Here is another example where I want to exclude ALL channels EXCEPT for a select few that match the exclusion list:

![image](https://github.com/euzu/m3u-filter/assets/160749165/1bb426cd-31f8-4d29-a3d3-a4dcd0db8b54)

Based on my experimentation and reading of the documentation all night this is the closest I got to recreating this in m3u-filter, but this is wrong.

![image](https://github.com/euzu/m3u-filter/assets/160749165/6978dea6-05bd-45fc-b692-ff5d3be27cc8)

Could you help me understand the best way to accomplish my goal? This is what I want to do in a nutshell:

1. Input very large xtream API
2. Select many groups (sports is just one of many) and filter based on certain titles
3. export the results to a new xtream API

I believe that's what you wrote this project for and I'd love to get rid of Threadfin since it doesn't expose an xtream api (which I need).

**Comment by euzu** [comment]
@RohirrimRider  this exactly why i have started this project.

I will try to explain how you can achieve this.

Tomorrow I try to write down a scenario and the steps to solve it.

**Reply by zero-well** [reply]
@euzu thank you so much!

**Reply by euzu** [reply, marked as answer]
I have written a howto for the initial setup of m3u-fitler for xtream provider.

[initial setup xtream provider](https://github.com/euzu/m3u-filter/wiki/scenario_xtream_provider)

**Reply by zero-well** [reply]
This is great!

**Comment by zero-well** [comment]
Okay @euzu I think I figured it out. tell me if this is wrong:

I have a bunch of these templates
```
templates:
  - name: ALL_CHANNELS
    value: >-
      (!Sports!)
      OR (!BallySports!)
      OR (!ESPN!)
      OR (!NetworkTv!)
      OR (!NCAAF!)


  - name: Sports
    value: >-
      Group ~ "Sports Networks"
      AND (NOT Title ~  ".*(Australia|TENNIS|BEIN|Bee |Cricket|TNT|SKY).*")

  - name: BallySports
    value: >-
      (Group ~ "Bally Sports")

  - name: NCAAF
    value: >-
      Group ~ "NCAAF"
      AND Title ~ "NCAAF \d+: .+"

  - name: ESPN
    value: >-
      (Group ~ "ESPN Play US"
      AND Title ~ "(?i)^\(US\) ESPN PLAY \d+: .+")
      OR (Group ~ "ESPN Plus.*"
      AND Title ~ "(?i)^ESPN\+ \d+: .+")

...
```

And then in my targets I have this:

```
      - name: with_locals
        output:
          - type: xtream
        filter: >-
          (!ALL_CHANNELS!) OR (!Locals!)
        mapping:
          - espn
          - 24/7
          - locals
          - NetworkTv
        sort:
          groups:
            order: asc
```

**Reply by euzu** [reply]
@RohirrimRider  yes this is exactly how it works.
You could write the complete filter definition to the `filter` property of the target, but it would lead to complex filter which you cant read and maintain. Templates are for splitting up in smaller peaces, for better readability, usability and maintainability.
If your regex are fine ,then everything should ok.
If you have some chars which are not on your keyboard, you can use the regex  like dot to match these.

You removed the `rename`  block. Thats fine because `mapping`  can rename too and can do more. The `rename` block is only for simple renames.

The default processing order  is filter-rename-map
You can change this, but be aware, if you first renamr or map, you need to adjust your filter to the new names. So for the start keep the default processing order.

Another point is  persist_file for xtream inputs. It is useless, because it is only for m3u targets.  The name for the xtream files target folder is the target name.

I dont.know how fit you are with regular expresssions. There is  regex101.com  where you can test regular expressions. 
Dont forget to change Flavour to Rust on the left side.
